url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/XMLDecoder.java

```diff
@@ -0,0 +1,705 @@
+            return exp.getValue();
+        }
+
+        private Method findMethod(Class<?> clazz, String methodName,
+                Class<?>[] clazzes) throws Exception {
+            Method[] methods = clazz.getMethods();
+            ArrayList<Method> matchMethods = new ArrayList<Method>();
+
+            // Add all matching methods into a ArrayList
+            for (Method method : methods) {
+                if (!methodName.equals(method.getName())) {
+                    continue;
+                }
+                Class<?>[] parameterTypes = method.getParameterTypes();
+                if (parameterTypes.length != clazzes.length) {
+                    continue;
+                }
+                boolean match = true;
+                for (int i = 0; i < parameterTypes.length; i++) {
+                    boolean isNull = (clazzes[i] == null);
+                    boolean isPrimitive = isPrimitiveWrapper(clazzes[i], parameterTypes[i]);
+                    boolean isAssignable = isNull? false : parameterTypes[i].isAssignableFrom(clazzes[i]);
+                    if ( isNull || isPrimitive || isAssignable ) {
+                        continue;
+                    }
+                    match = false;
+                }
+                if (match) {
+                    matchMethods.add(method);
+                }
+            }
+
+            int size = matchMethods.size();
+            if (size == 1) {
+                // Only one method matches, just invoke it
+                return matchMethods.get(0);
+            } else if (size == 0) {
+                // Does not find any matching one, throw exception
+                throw new NoSuchMethodException(Messages.getString(
+                        "custom.beans.41", methodName)); //$NON-NLS-1$
+            }
+
+            // There are more than one method matching the signature
+            // Find the most specific one to invoke
+            MethodComparator comparator = new MethodComparator(methodName,
+                    clazzes);
+            Method chosenOne = matchMethods.get(0);
+            matchMethods.remove(0);
+            int methodCounter = 1;
+            for (Method method : matchMethods) {
+                int difference = comparator.compare(chosenOne, method);
+                if (difference > 0) {
+                    chosenOne = method;
+                    methodCounter = 1;
+                } else if (difference == 0) {
+                    methodCounter++;
+                }
+            }
+            if (methodCounter > 1) {
+                // if 2 methods have same relevance, throw exception
+                throw new NoSuchMethodException(Messages.getString(
+                                                                   "custom.beans.62", methodName)); //$NON-NLS-1$
+            }
+            return chosenOne;
+        }
+
+        private boolean isPrimitiveWrapper(Class<?> wrapper, Class<?> base) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 456-463 into a separate method.